### PR TITLE
feat: enhance IsValid() to validate value type compatibility

### LINKED ISSUE #293: Inconsistent Exception Handling for Invalid Filter Values Across Different Types
### Version

v2.17.0

## Problem Description

When invalid values are used in filters, Gridify exhibits inconsistent behavior across different data types:

- **For `int` properties:** Throws `ArgumentException`
- **For `DateTime` properties:** Returns empty results (no exception)
- **For `Enum` properties:** Returns empty results (no exception)

This inconsistency creates poor user experience and makes error handling unpredictable.

## Steps to Reproduce
```csharp
// Setup
public class TestEntity
{
    public int MyInt { get; set; }
    public DateTime MyDate { get; set; }
    public MyEnum MyEnum { get; set; }
}

var data = GetTestData().AsQueryable();

// Test 1: DateTime with invalid value
var query1 = new GridifyQuery { Filter = "MyDate=xyz" };
var result1 = data.Gridify(query1);
// Result: Returns empty array [] (no exception)

// Test 2: Enum with invalid value  
var query2 = new GridifyQuery { Filter = "MyEnum>999999999" };
var result2 = data.Gridify(query2);
// Result: Returns empty array [] (no exception)

// Test 3: Int with invalid value
var query3 = new GridifyQuery { Filter = "MyInt=xyz" };
var result3 = data.Gridify(query3);
// Result: Throws Exception!
// Exception: "Object of type 'System.String' cannot be converted to type 'System.Int32'"
```

## Expected Behavior

All invalid values should be handled consistently - either:
1. All return empty results (preferred for better UX), OR
2. All throw meaningful exceptions

Additionally, the `IsValid()` method should validate value types, not just field names.

## Actual Behavior

Mixed behavior depending on which exception type is thrown during type conversion.

### Steps to reproduce

1. Create entity with `int`, `DateTime`, and `Enum` properties
2. Try filtering with invalid string values for each type
3. Observe different behaviors:
   - `int` throws exception
   - `DateTime` and `Enum` return empty results

### PR DESCRIPTION
**Pull request by yasserebrahimi**
##  Description
This PR enhances the `IsValid()` method to validate value type compatibility, addressing issue #293.

##  Problem
Currently, `IsValid()` only checks if field names are mapped but doesn't validate whether filter values can be converted to the target property types. This causes inconsistent behavior:
- **Integer fields**: Throws exception at runtime with invalid values
- **DateTime fields**: Returns empty results silently  
- **Enum fields**: Returns empty results silently

This inconsistency makes it difficult for users to provide meaningful validation feedback.

##  Solution
Added a new overload: `IsValid<T>(out List<string> validationErrors)` that:
- ✅ Validates field name mappings (existing behavior)
- ✅ Validates value type compatibility for all property types
- ✅ Returns detailed error messages for each validation failure
- ✅ Maintains full backward compatibility

##  Changes
- Added `IsValid<T>(out List<string> validationErrors)` overload
- Added `ExtractPropertyType()` helper using reflection to get actual property types from expression trees
- Added `FindValueExpressionsForField()` and `TraverseForValues()` for syntax tree traversal
- Added `TryConvertValue()` with special handling for bool, Guid, and TypeConverter-based types
- Added 25+ comprehensive unit tests covering all scenarios

##  Testing
All tests pass locally:
```bash
dotnet test -c Release
# Total: 25+ new tests, all passing
```

**Test coverage includes:**
- ✅ Valid values for all types (int, DateTime, Enum, bool, Guid, decimal, etc.)
- ✅ Invalid values for all types with proper error messages
- ✅ Edge cases (null, empty, overflow)
- ✅ Complex queries with multiple conditions
- ✅ Custom mappers
- ✅ Nullable types
- ✅ Different operators (=, !=, >, <, >=, <=)
- ✅ Backward compatibility with existing `IsValid()` method

##  Example Usage
```csharp
var query = new GridifyQuery { Filter = "Age=abc" };

// New overload with detailed errors
var isValid = query.IsValid(out var errors);
// isValid: false
// errors: ["Cannot convert value 'abc' to type 'Int32' for field 'Age': Invalid format"]

// Old overload still works
var isValid = query.IsValid();
// isValid: false
```

##  Breaking Changes
**None** - This is a non-breaking enhancement for v2.
- Existing `IsValid()` method unchanged
- New overload is opt-in
- All existing code continues to work

##  Checklist
- [x] Code follows project style guidelines
- [x] All tests pass
- [x] Added comprehensive unit tests
- [x] Backward compatible
- [x] Professional English comments
- [x] XML documentation added

##  Fixes
Closes #293

---

Thank you for considering this contribution! I'm happy to make any changes based on your feedback. 🙏

### CODE REVIEW

File: src/Gridify/GridifyExtensions.cs [hunk 1]
```diff
@@ -483,4 +485,243 @@ public static Paging<T> Gridify<T>(this IQueryable<T> query, Action<IGridifyQuer
...
    #endregion
+
+   /// <summary>
+   /// Validates the GridifyQuery including field names and value type compatibility.
+   /// </summary>
+   /// <param name="filtering">The filtering query to validate.</param>
+   /// <param name="validationErrors">List of validation error messages if validation fails.</param>
+   /// <param name="mapper">Optional custom mapper.</param>
+   /// <returns>True if the query is valid; otherwise, false.</returns>
+   public static bool IsValid<T>(
```
Comment (alirezanet) [inline src/Gridify/GridifyExtensions.cs]: This is actually a great idea but we need to have a consistent behavior between these two overloads, right now this overload complains about some mismatched value types but the other one return true

I would suggest to call this overload in the old one, and discard the `out` parameter `_`.

File: test/Gridify.Tests/IssueTests/Issue293Tests.cs [hunk 2]
```diff
@@ -0,0 +1,299 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace Gridify.Tests.IssueTests;
+
+public class Issue293Tests
```
Comment (alirezanet) [inline test/Gridify.Tests/IssueTests/Issue293Tests.cs]: You're convering more than cases than just the issue related ones, Please move this class out of the issues folder and name it `GridifyQueryTests` or `GridifyQueryValidationTests`

**Review by alirezanet — CHANGES_REQUESTED** [review]
Hi Yasser,
Thanks for your contribution,💐 
Overall it looks good but we still need a few tweaks: 

1- Consistent behavior over all `IsValid` overloads
2- Tests for the old IsValid method
3- We need to also update the documentation regarding the new overload

**Review by alirezanet — APPROVED** [review]
Great Job 🎉👏, LGTM. 

> The manual traversal in FindValueExpressionsForField() and TraverseForValues() builds on the existing field extraction but adds the value-to-field association needed for type conversion validation. 
> I'm happy to refactor if there's a better approach using existing Gridify utilities!


There’s still some room to reduce tree traversal, we currently do it twice: once to extract the fields and once to get the values by field name. I **don’t** see this as a blocker though; any performance gain would be minor.
image


I’ve already approved it, but if you have time, please update the [documentation](https://alirezanet.github.io/Gridify/guide/gridifyQuery#isvalid) with the new overload. Thanks!
